tests/functional/test_main.py

```diff
@@ -0,0 +1,229 @@
+
+
+def test_user_detail_page_404(client, app):
+
+    response = client.get(f"/users/not_a_real_user")
+    assert b"404" in response.data
+
+
+def test_user_search_no_results(client, app, persisted_user):
+    response = client.get("/usersearch", query_string={"q": "fakeuser"})
+    assert b"no results found" in response.data
+
+
+def test_song_search(client, app):
+    response = client.get("/search")
+    assert b"Search for a song title" in response.data
+
+
+def test_song_search_post(client, app):
+    response = client.post("/search", data={"search": "song_title"})
+    assert response.status_code == HTTPStatus.FOUND
+
+
+def test_edit_user_get(app, test_db, persisted_user):
+    """
+    Given a registered user
+    When that user attempts to view the edit-profile page
+    Then the page should be found without a 404
+    """
+    with app.test_client(user=persisted_user) as test_client:
+        response = test_client.get("/edit-profile")
+        current_user = persisted_user
+
+    assert response.status_code == HTTPStatus.OK
+    assert current_user.id == persisted_user.id
+    assert bytes(current_user.email, "utf8") in response.data
+    assert b"404 " not in response.data
+    assert b"503 " not in response.data
+
+
+def test_user_search(client, app):
+    response = client.get("/usersearch")
+
+    assert b"Search for a user" in response.data
+    assert response.status_code == HTTPStatus.OK
+
+
+def test_user_search_with_query(client, app, persisted_user):
+    response = client.get("/usersearch?q=test")
+
+    assert bytes(persisted_user.username, "utf8") in response.data
+    assert response.status_code == HTTPStatus.OK
+
+
+def test_user_search_with_bad_query(client, app, persisted_user):
+    response = client.get("/usersearch?q=jibberish")
+
+    assert b"Sorry, no results found" in response.data
+    assert response.status_code == HTTPStatus.OK
+
+
+def test_song_search(app, test_db, persisted_user):
+    with app.test_client(user=persisted_user) as test_client:
+        response = test_client.get("/search")
+        current_user = persisted_user
+
+    assert response.status_code == HTTPStatus.OK
+    assert b"Search for a song title" in response.data
+
+
+def test_song_search_results(client, app, mocker):
+    """Not sure if mock implemented properly?"""
+    mocker.patch(
+        "app.views.LastFMSearch.get_results",
+        return_value={
+            "result": {
+                "results": {
+                    "trackmatches": {
+                        "track": [
+                            {
+                                "name": "Army of Me",
+                                "artist": "björk",
+                                "url": "https://www.last.fm/music/Bj%C3%B6rk/_/Army+of+Me",
+                                "image": "/static/user_icon-01.png"
+                            }
+                        ]
+                    }
+                }
+            }
+        },
+    )
```

Comment:
this could be a fixture called mocked_lastFM_search that returns the return value in case you need data from that dict